File: javascript/p004.js
/*
 * Javascript Solution to Project Euler problem 4
 * by Mychal Johnson
 *
 * https://github.com/mychalj/project-euler
 */

//Determines whether the number is a palidrome.
function isPalindrome(num) {
  if (num === parseInt(num.toString().split("").reverse().join(""))) {
    return true;
  } else {
    return false;
  }
}

//Calculates the largest palindrome that was requested.
function largest_palindrome(min, max) {
  var i, j, product, answer = 0;

  for (i = min; i < max; i++) {
    for (j = i; j < max; j++) {
      product = i * j;
      if (product > answer && isPalindrome(product)) { answer = product; }
    }
  }
  return answer
}

//Outputs the largest palindrome.
console.log(largest_palindrome(100, 999));

File: ruby/p004.rb
# Ruby Solution to Project Euler problem 4
# by Mychal Johnson
# https://github.com/mychalj/project-euler

# Determines whether the number is a palidrome.
def palindrome?(num)
  return true if (n = num.to_s) == n.reverse
  false
end

# Calculates the largest palindrome that was requested.
def largest_palindrome(min, max)
  range = (min..max)
  max = range.end
  answer = 0

  range.each do |i|
    max.downto(i) do |j|
      product = i * j
      answer = product if product > answer && palindrome?(product)
    end
  end
  puts answer
end

largest_palindrome(100, 999)
